File: sistema_cpluplus/codigos/PagamentoPaciente.cpp
#include "PagamentoPaciente.h"
#include <iostream>//entrada e saida


using namespace std;

//construtor
PagamentoPaciente :: PagamentoPaciente() {}


//construtor
PagamentoPaciente::PagamentoPaciente(Paciente& paciente)
{
	this->paciente = paciente;
}

//construtor
PagamentoPaciente::PagamentoPaciente(PagamentoPaciente& pagamento)
{
	this->paciente = pagamento.paciente;

	//inserir novos elementos
	for (auto it = pagamento.pagamentosPaciente.begin(); it != pagamento.pagamentosPaciente.end(); it++)
	{
		this->pagamentosPaciente.push_back(new ItemPagamentoPaciente(**it));
	}
	
}

//destrutor

PagamentoPaciente:: ~PagamentoPaciente() 
{
	for (auto item : this->pagamentosPaciente)
	{
		if (item != nullptr)
		{
			delete item;
		}
	}
}

void PagamentoPaciente:: adicionarItemPagamentoPaciente(ItemPagamentoPaciente item)//adiciona um ItemPago pelo paciente de maneira ordenada pela data
{
	//inserir novos elementos
	ItemPagamentoPaciente* copia = new ItemPagamentoPaciente(item);

	list < ItemPagamentoPaciente*> ::iterator it = this->pagamentosPaciente.begin();

	for (; it != this->pagamentosPaciente.end(); it++)
	{
		if ((**it).dataDePagamento > (*copia).dataDePagamento)//se o ano a inserir for menor ja pode inserir
		{
			break;
		}
			
		else if ((*copia).dataDePagamento > (**it).dataDePagamento)//se for maior, pode continuar a iterar ate achar o ano
		{
			continue;
		}

		else//eh igual
		{
			break;
		}

	}

	this->pagamentosPaciente.insert(it, copia);
}

bool PagamentoPaciente:: removerItemPagamentoPaciente(int posicao)//dada uma posicao remove um item pago pelo paciente
{
	if (pagamentosPaciente.empty() || posicao > pagamentosPaciente.size() || posicao < 1)
	{
		return false;
	}
	int i = 0;
	for (auto it = pagamentosPaciente.begin(); it != pagamentosPaciente.end(); it++)
	{
		if (i == posicao - 1)
		{
			delete *it;
			pagamentosPaciente.erase(it);
			break;
		}
		i++;
	}

	return true;
}

ItemPagamentoPaciente& PagamentoPaciente :: retornarItemPagamentoPaciente(int posicao)//retorna um item pago pelo paciente
{
	int i = 0;

	auto it = pagamentosPaciente.begin();
	for (; it != pagamentosPaciente.end(); it++)
	{
		if (i == posicao - 1)
		{
			
			return **it;
		}
		i++;
	}

	return **it;///retorna ultimo caso nao encontre
}

bool PagamentoPaciente::imprimirItensPagosNoMes(int mes)
{
	int i = 0;

	if (pagamentosPaciente.empty() || mes < 0 || mes > 12)
	{
		return false;
	}

	for (auto it = pagamentosPaciente.begin(); it != pagamentosPaciente.end(); it++)
	{
		
		if ((**it).dataDePagamento.getMes() == mes)
		{
			i++;
			cout << "Pagamento " << i << ": " << (**it) << "\n";

		}
	}

	return i;
}

//sobrecarga do operador de atribuicao '=' 

PagamentoPaciente& PagamentoPaciente:: operator=(PagamentoPaciente& pagamento)
{
	if (this == &pagamento)
	{
		return *this;
	}

	this->paciente = pagamento.paciente;
	
	//desalocar lista
	for (auto desalocar : this->pagamentosPaciente)
	{
		if (desalocar != nullptr)
		{
			delete desalocar;
		}
	}

	this->pagamentosPaciente.clear();//limpar a lista 

	//inserir novos elementos
	for (auto it = pagamento.pagamentosPaciente.begin(); it != pagamento.pagamentosPaciente.end(); it++)
	{
		this->pagamentosPaciente.push_back(new ItemPagamentoPaciente(**it));
	}

	return *this;
}


//sobrecarga do operador 'cout'
ostream& operator << (ostream& out, PagamentoPaciente& pagamento)
{
	out << "Paciente: " << pagamento.paciente.getNome() << " CPF:" << pagamento.paciente.getCpf() << "\n\n";

	if (pagamento.pagamentosPaciente.empty())//se nao tiver pagamentos
	{
		out << "Sem pagamentos";
	}
	else//caso contrario imprime os pagamentos
	{
		int i = 1;
		for (auto it = pagamento.pagamentosPaciente.begin(); it != pagamento.pagamentosPaciente.end(); it++)
		{
			out << "Pagamento " << i << ": " << "\n" << **it <<"\n";
			i++;
		}
	}

	return out;
}

File: sistema_java/codigos/oficinapoo/PagamentoPaciente.java
package oficinapoo;

import java.util.List;
import java.util.LinkedList;
import java.util.Collections;


public class PagamentoPaciente
{
    public Paciente paciente;
    public List<ItemPagamentoPaciente> pagamentosPaciente;
    
    //construtor
    public PagamentoPaciente()
    {
        paciente = new Paciente();
        pagamentosPaciente = new LinkedList<ItemPagamentoPaciente>();
    }


    //construtor
    public PagamentoPaciente(Paciente paciente)
    {
        this.paciente = new Paciente(paciente);
        pagamentosPaciente = new LinkedList<ItemPagamentoPaciente>();
    }   

    //construtor
    public PagamentoPaciente(PagamentoPaciente pagamento)
    {
        this.paciente = new Paciente(pagamento.paciente);

	//inserir novos elementos
        this.pagamentosPaciente = new LinkedList<ItemPagamentoPaciente>();
        
        for(ItemPagamentoPaciente it : pagamento.pagamentosPaciente)
        {
            this.pagamentosPaciente.add(new ItemPagamentoPaciente(it));
        }	
    }
    
     //ordena a os item pagos pelo paciente pela data
    public void ordenaPagamentosPaciente()
    {
        Collections.sort(pagamentosPaciente);                                                             
    }

    public void adicionarItemPagamentoPaciente(ItemPagamentoPaciente item)//adiciona um ItemPago pelo paciente de maneira ordenada pela data
    {
	//inserir novos elementos
	ItemPagamentoPaciente copia = new ItemPagamentoPaciente(item);
       
        //adicionar o pagamento do paciente
	this.pagamentosPaciente.add(copia);
        
        //ordenar a lista
        ordenaPagamentosPaciente();     
    }
    

    public boolean removerItemPagamentoPaciente(int posicao)//dada uma posicao remove um item pago pelo paciente
    {
	//verifica se posicao eh valida e se lista de pagamentos de pacientes nao eh vazia
	if (pagamentosPaciente.isEmpty() || posicao >  pagamentosPaciente.size() || posicao < 1)
	{
            return false;
	}
	
        pagamentosPaciente.remove(posicao-1);
        
	return true;
    }

    public ItemPagamentoPaciente retornarItemPagamentoPaciente(int posicao)//retorna um item pago pelo paciente
    {
	//se a posicao for incorreta retorna o ultimmo
        if(posicao > pagamentosPaciente.size() || posicao < 1)
        {
            return pagamentosPaciente.get(pagamentosPaciente.size()-1);
        }
        
	return pagamentosPaciente.get(posicao-1);
    }

    public String imprimirItensPagosNoMes(int mes)//imprime os itens pagos no mes pelo funcionario
    {
        String leitura = "";
	int i = 0;

        //se a lista estiver vazia ou mes incorreto retorna falso
	if (pagamentosPaciente.isEmpty() || mes < 0 || mes > 12)
	{
		return "Sem Itens";
	}

	for (ItemPagamentoPaciente it : pagamentosPaciente)
	{
            if (it.dataDePagamento.getMes() == mes)
            {
		i++;
		leitura += "Pagamento "+i+": "+it+"\n";
            }
	}

        return leitura;
	//return i;
    }
    
     //imprime os Pagamentos do Paciente
    public String toString()
    {
        String retorno = "";
        
        retorno = retorno + "Paciente: " + paciente.getNome() + " CPF:" + paciente.getCpf() + "\n\n";
        
        if ( pagamentosPaciente.isEmpty())//se nao tiver itens de Pagamentos
	{
		retorno = retorno+ "Sem Pagamentos";
	}
        else
        {
            int i = 1;
            for (ItemPagamentoPaciente it : pagamentosPaciente)
            {
		retorno = retorno + "Pagamento " + i + ": " + "\n" + it.toString() + "\n";
		i++;
            }
        }
        
        return retorno;
    }  
	
}
